src/lib/openssl_preload/crypto.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * The Shadow Simulator
+ * Copyright (c) 2010-2011, Rob Jansen
+ * See LICENSE for licensing information
+ */
+
+#define _GNU_SOURCE
+#include <dlfcn.h>
+#include <execinfo.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+// Caches whether or not a memory address is from libssl.so
+typedef struct {
+    void* addr;
+    bool is_libssl;
+} bt_cache_entry_t;
+
+// We use a small cache size: in ad-hoc experiments with tor-0.4.6.9, we observed
+// at most three callers of EVP_EncryptUpdate.
+#define EVP_BACKTRACE_CACHE_LEN 10
+static bt_cache_entry_t evp_backtrace_cache[EVP_BACKTRACE_CACHE_LEN];
+
+// For storing a pointer to the original EVP_EncryptUpdate function.
+typedef int EVP_EncryptUpdate_func(void*, unsigned char*, int*, const unsigned char*, int);
+static void* evp_eu_funcptr = NULL;
+
+#ifdef DEBUG
+// Counters for verifying that interception is happening correctly.
+static unsigned long aes_e_cnt = 0;
```

Comment:
These also aren't thread-safe.

We could either:
* Protect them with a mutex (maybe a single global mutex that also protects the cache)
* Document that this code is thread  unsafe except under Shadow
* Make these individually atomics  and use the corresponding operators like [atomic_fetch_add](https://en.cppreference.com/w/c/atomic/atomic_fetch_add). If we're going to make this code thread-safe though, I think the first option (single global mutex) would be more efficient assuming no contention in practice.

Reply:
Added a global mutex to protect global state.